### PR TITLE
New Base classes for Application and Configuration objects

### PR DESCRIPTION
**Pull request by clamdad**
- New `ConfigOptions` can be read/written from YAML and subclassed to add more config parameters.
- The `Application` base class defines
  - Shared Threading model with Startup and  Shutdown 
  - home directory
  - configuration options
  - logging
- `Application` is intended for subclassing and customization by core components including
  - `ReporterApplication`
  - `TellormonApplication`
  - `FeedDatabaseApplication`
 
For usage, see
- `tests/test_config.py`
- `tests/test_app.py`

### CODE REVIEW

File: src/telliot/utils/config.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        """Load Configuration from a .yaml file
+
+        Parameters
+        ----------
+        filename
+        """
+        with open(filename, "r") as f:
+            state = yaml.load(f, Loader=Loader)
+
+        return cls.parse_obj(state)
```
Comment (oraclown) [inline src/telliot/utils/config.py:45]: What is `cls`?

File: src/telliot/utils/config.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import pathlib
+from typing import Union
+
+import yaml
+from pydantic import BaseModel
+from yaml import CDumper as Dumper
+from yaml import CLoader as Loader
+
+
+class LogLevel(str, enum.Enum):
```
Comment (oraclown) [inline src/telliot/utils/config.py:11]: Why do we have to enumerate logging levels if Python's built-in loggers already let you use different logging levels like: `logging.error("ahhh")` or `logging.info("hello")`?

**Review by oraclown — APPROVED** [review]
Looks good! Just had a couple questions that I left as comments.

File: src/telliot/utils/config.py (hunk 1 above)
Comment (clamdad) [reply src/telliot/utils/config.py:45]: Here, `cls` is the usual class object passed in through a[`classmethod`](https://docs.python.org/3/library/functions.html#classmethod) constructor

File: src/telliot/utils/config.py (hunk 2 above)
Comment (clamdad) [reply src/telliot/utils/config.py:11]: The idea is for the configuration to allow the user to choose the level of logging that he/she would like to see in the output.

However, I am not thrilled about having to enumerate the possible values in the `class LogLevel(str, enum.Enum)` class.  I was just trying to manipulate pydantic to export/import from a string in the config file.  Definitely open to improvements/suggestions here.

**Review by tallywiesenberg — APPROVED** [review]
Looks good to me, and thanks @oraclown for the questions

File: src/telliot/utils/config.py (hunk 2 above)
Comment (clamdad) [reply src/telliot/utils/config.py:11]: I'm going to go ahead and merge this change,  but feel free to improve on the enumeration in the future.

**Comment by clamdad** [comment]
Thanks for the feedback, gents.  I'll go ahead and merge.

**Pull request merged**